<a href="https://colab.research.google.com/github/KleverGH/DATA_MINING/blob/main/601_TF_Limpieza_de_datos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns

In [ ]:
df = pd.read_csv("Dataset/Enaho01-2022-601.csv", encoding='ISO-8859-1')

C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\3849066368.py:1: DtypeWarning: Columns (34) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("Dataset/Enaho01-2022-601.csv", encoding='ISO-8859-1')


In [ ]:
df.head()

AÑO  MES  CONGLOME  VIVIENDA  HOGAR  UBIGEO  DOMINIO  ESTRATO  P601A  \
0  2022    1      5030         8     11   10201        7        4    100   
1  2022    1      5030         8     11   10201        7        4    101   
2  2022    1      5030         8     11   10201        7        4    102   
3  2022    1      5030         8     11   10201        7        4    103   
4  2022    1      5030         8     11   10201        7        4    104   

                       P601X  ...             D601C BLIBRE22 T601A71 T601A72  \
0  PAN FRANCES Y OTROS PANES  ...                                  0       0   
1      Pan Corriente o común  ...  194.898147583008                0       0   
2                 Pan Tolete  ...                                              
3                Pan Frances  ...                                              
4                 Pan Chapla  ...                                              

  PRODUC61 TICUEST01 I601E    FACTOR07 NCONGLOME SUB_CONGLOME  
0        0         2        106.890244      6618            0  
1   110091         2        106.890244      6618            0  
2   110076         2        106.890244      6618            0  
3   110074         2        106.890244      6618            0  
4   110093         2        106.890244      6618            0  

[5 rows x 39 columns]

In [ ]:
df_filtered = df[df['P601B'] == 1]

Se aplica este filtro en la columna P601B debido a que esta columna representa la pregunta si es que en los ultimos 15 dias ha realizado alguna compra, entonces cuando el resultado es NO, se coloca 2, lo que genera que todo el data set para ese registro (cuando es igual a 2) tenga valores nulos o vacios. Esto es un claro ejemplo de una forma en la que se puede reducir el data set.

In [ ]:
df_filtered.shape

(2733695, 39)

In [ ]:
range = ['P601A1','P601A2','P601A3','P601A4','P601A5','P601A6','P601A7']

In [ ]:
for columna in range:
    df_filtered[columna].replace(" ","0",inplace=True)

C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\3809193040.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered[columna].replace(" ","0",inplace=True)


In [ ]:
for columna in range:
    print("Valores unicos para ",columna," es: ",df_filtered[columna].unique(),'\n')

Valores unicos para  P601A1  es:  ['1' '0'] 

Valores unicos para  P601A2  es:  ['0' '1'] 

Valores unicos para  P601A3  es:  ['0' '1'] 

Valores unicos para  P601A4  es:  ['0' '1'] 

Valores unicos para  P601A5  es:  ['0' '1'] 

Valores unicos para  P601A6  es:  ['0' '1'] 

Valores unicos para  P601A7  es:  ['0' '1'] 



In [ ]:
mapeo = {
    'P601A1': 'Comprado',
    'P601A2': 'Autoconsumo',
    'P601A3': 'Autosuministro',
    'P601A4': 'Como parte de pago a un miembro del hogar',
    'P601A5': 'Regalado o pagado por algún miembro de otro hogar',
    'P601A6': 'Regalado o donado por algún programa social',
    'P601A7': 'Otro'
}

In [ ]:
# Convertir las columnas dummy a tipo int
for col in mapeo.keys():
    df_filtered[col] = df_filtered[col].astype(int)

# Luego, aplica el método idxmax
df_filtered['Forma de obtención de producto'] = df_filtered[list(mapeo.keys())].idxmax(axis=1)
df_filtered['Forma de obtención de producto'] = df_filtered['Forma de obtención de producto'].map(mapeo)

C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\4257039407.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered[col] = df_filtered[col].astype(int)
C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\4257039407.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered['Forma de obtención de producto'] = df_filtered[list(mapeo.keys())].idxmax(axis=1)
C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\4257039407.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice fr

In [ ]:
df_filtered['Forma de obtención de producto'].unique()

array(['Comprado', 'Otro', 'Autosuministro',
       'Regalado o pagado por algún miembro de otro hogar', 'Autoconsumo',
       'Regalado o donado por algún programa social',
       'Como parte de pago a un miembro del hogar'], dtype=object)

In [ ]:
df_filtered['Forma de obtención de producto'].value_counts()

Comprado                                             2238490
Autoconsumo                                           153281
Regalado o pagado por algún miembro de otro hogar     140293
Autosuministro                                         83562
Otro                                                   77414
Regalado o donado por algún programa social            37736
Como parte de pago a un miembro del hogar               2919
Name: Forma de obtención de producto, dtype: int64

In [ ]:
df_filtered = df_filtered.drop(columns=range)

In [ ]:
df_filtered.head()

AÑO  MES  CONGLOME  VIVIENDA  HOGAR  UBIGEO  DOMINIO  ESTRATO  P601A  \
0   2022    1      5030         8     11   10201        7        4    100   
1   2022    1      5030         8     11   10201        7        4    101   
8   2022    1      5030         8     11   10201        7        4    108   
15  2022    1      5030         8     11   10201        7        4    300   
16  2022    1      5030         8     11   10201        7        4    301   

                        P601X  ...  BLIBRE22 T601A71 T601A72 PRODUC61  \
0   PAN FRANCES Y OTROS PANES  ...                 0       0        0   
1       Pan Corriente o común  ...                 0       0   110091   
8   TOSTADAS INTEGRALES A GRA  ...                 0       0   110087   
15  ARROZ (CORRIENTE Y SUPERI  ...                 0       0        0   
16   Arroz Corriente a Granel  ...                 0       0   110100   

   TICUEST01 I601E    FACTOR07 NCONGLOME SUB_CONGLOME  \
0          2        106.890244      6618            0   
1          2        106.890244      6618            0   
8          2        106.890244      6618            0   
15         2        106.890244      6618            0   
16         2        106.890244      6618            0   

   Forma de obtención de producto  
0                        Comprado  
1                        Comprado  
8                        Comprado  
15                       Comprado  
16                       Comprado  

[5 rows x 33 columns]

In [ ]:
columns_to_select = ['CONGLOME', 'VIVIENDA', 'HOGAR', 'P601X', 'Forma de obtención de producto', 'P601B1', 'P601B2', 'P601B3', 'P601B4', 'PRODUC61']
df_selected = df_filtered[columns_to_select]

In [ ]:
df_selected.head()

CONGLOME  VIVIENDA  HOGAR                      P601X  \
0       5030         8     11  PAN FRANCES Y OTROS PANES   
1       5030         8     11      Pan Corriente o común   
8       5030         8     11  TOSTADAS INTEGRALES A GRA   
15      5030         8     11  ARROZ (CORRIENTE Y SUPERI   
16      5030         8     11   Arroz Corriente a Granel   

   Forma de obtención de producto P601B1            P601B2 P601B3 P601B4  \
0                        Comprado                                          
1                        Comprado      2               150      2      6   
8                        Comprado      4  .370000004768372      9      4   
15                       Comprado                                          
16                       Comprado      2                 1      9      6   

    PRODUC61  
0          0  
1     110091  
8     110087  
15         0  
16    110100

In [ ]:
df_selected.shape

(2733695, 10)

In [ ]:
df_selected = df_selected.rename(columns={
    'P601X': 'Nombre del producto',
    'P601B1': 'Frecuencia de compra',
    'P601B2': 'Precio',
    'P601B3': 'Cantidad comprada (unidad de medida)',
    'P601B4': 'Lugar de compra',
    'PRODUC61': 'Codigo de producto'
})

In [ ]:
df_selected.head()

CONGLOME  VIVIENDA  HOGAR        Nombre del producto  \
0       5030         8     11  PAN FRANCES Y OTROS PANES   
1       5030         8     11      Pan Corriente o común   
8       5030         8     11  TOSTADAS INTEGRALES A GRA   
15      5030         8     11  ARROZ (CORRIENTE Y SUPERI   
16      5030         8     11   Arroz Corriente a Granel   

   Forma de obtención de producto Frecuencia de compra            Precio  \
0                        Comprado                                          
1                        Comprado                    2               150   
8                        Comprado                    4  .370000004768372   
15                       Comprado                                          
16                       Comprado                    2                 1   

   Cantidad comprada (unidad de medida) Lugar de compra  Codigo de producto  
0                                                                         0  
1                                     2               6              110091  
8                                     9               4              110087  
15                                                                        0  
16                                    9               6              110100

In [ ]:
df_final = df_selected[df_selected['Codigo de producto'] != 0]

Se excluyen todos los registros en que los productos son iguales a 0 debido a que presentan nulos en todas las variables a comparar, asimismo, no se puede realizar un reemplazo a los valores vacios debido a que el procedimiento que se plantea realizar es un reemplazo por la moda en base al codigo de producto, pero como el codigo de producto es 0, no hay valores con que reemplazar.

In [ ]:
df_final.shape

(1759681, 10)

In [ ]:
df_final.head()

CONGLOME  VIVIENDA  HOGAR        Nombre del producto  \
1       5030         8     11      Pan Corriente o común   
8       5030         8     11  TOSTADAS INTEGRALES A GRA   
16      5030         8     11   Arroz Corriente a Granel   
28      5030         8     11                Papa Blanca   
29      5030         8     11              Papa Amarilla   

   Forma de obtención de producto Frecuencia de compra            Precio  \
1                        Comprado                    2               150   
8                        Comprado                    4  .370000004768372   
16                       Comprado                    2                 1   
28                       Comprado                    3                 1   
29                       Comprado                    3                .5   

   Cantidad comprada (unidad de medida) Lugar de compra  Codigo de producto  
1                                     2               6              110091  
8                                     9               4              110087  
16                                    9               6              110100  
28                                    9               6              110915  
29                                    9               6              110913

In [ ]:
# Reemplazar espacios en blanco con NaN
df_final['Precio'] = df_final['Precio'].replace(' ', np.nan)

# Convertir la columna a float
df_final['Precio'] = df_final['Precio'].astype(float)

# Opcional: reemplazar NaN con 0 (o cualquier otro valor que prefieras)
# df_final['Precio'] = df_final['Precio'].fillna(0)

# Redondear a 2 decimales
df_final['Precio'] = df_final['Precio'].round(2)

C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\4246219464.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['Precio'] = df_final['Precio'].replace(' ', np.nan)
C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\4246219464.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['Precio'] = df_final['Precio'].astype(float)
C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\4246219464.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .

In [ ]:
df_final.shape

(1759681, 10)

In [ ]:
# Contar valores nulos en cada columna
null_counts = df_final.isna().sum()
print("Cantidad de valores nulos por columna:")
print(null_counts)

# Contar valores vacíos (cadenas vacías) en cada columna
empty_string_counts = df_final.applymap(lambda x: x == '').sum()
print("\nCantidad de cadenas vacías por columna:")
print(empty_string_counts)

# Sumar ambos conteos para obtener el total de valores "ausentes" por columna
total_absent_values = null_counts + empty_string_counts
print("\nCantidad total de valores ausentes (nulos o cadenas vacías) por columna:")
print(total_absent_values)

Cantidad de valores nulos por columna:
CONGLOME                                     0
VIVIENDA                                     0
HOGAR                                        0
Nombre del producto                          0
Forma de obtención de producto               0
Frecuencia de compra                         0
Precio                                  350060
Cantidad comprada (unidad de medida)         0
Lugar de compra                              0
Codigo de producto                           0
dtype: int64

Cantidad de cadenas vacías por columna:
CONGLOME                                0
VIVIENDA                                0
HOGAR                                   0
Nombre del producto                     0
Forma de obtención de producto          0
Frecuencia de compra                    0
Precio                                  0
Cantidad comprada (unidad de medida)    0
Lugar de compra                         0
Codigo de producto                      0
dtype: int64

Can

In [ ]:
# Convertir valores vacíos en NaN
df_final['Precio'] = df_final['Precio'].replace('', np.nan).astype(float)

# Calcular el promedio de 'Precio' por 'Codigo de producto', excluyendo los valores nulos
avg_price_by_product = df_final.groupby('Codigo de producto')['Precio'].transform('mean')

# Reemplazar valores nulos en 'Precio' con el promedio calculado
df_final['Precio'].fillna(avg_price_by_product, inplace=True)

C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\458337301.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['Precio'] = df_final['Precio'].replace('', np.nan).astype(float)
C:\Users\dchir\AppData\Local\Temp\ipykernel_16328\458337301.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_final['Precio'].fillna(avg_price_by_product, inplace=True)


In [ ]:
# Contar valores nulos en cada columna
null_counts = df_final.isna().sum()
print("Cantidad de valores nulos por columna:")
print(null_counts)

# Contar valores vacíos (cadenas vacías) en cada columna
empty_string_counts = df_final.applymap(lambda x: x == '').sum()
print("\nCantidad de cadenas vacías por columna:")
print(empty_string_counts)

# Sumar ambos conteos para obtener el total de valores "ausentes" por columna
total_absent_values = null_counts + empty_string_counts
print("\nCantidad total de valores ausentes (nulos o cadenas vacías) por columna:")
print(total_absent_values)

Cantidad de valores nulos por columna:
CONGLOME                                  0
VIVIENDA                                  0
HOGAR                                     0
Nombre del producto                       0
Forma de obtención de producto            0
Frecuencia de compra                      0
Precio                                  522
Cantidad comprada (unidad de medida)      0
Lugar de compra                           0
Codigo de producto                        0
dtype: int64

Cantidad de cadenas vacías por columna:
CONGLOME                                0
VIVIENDA                                0
HOGAR                                   0
Nombre del producto                     0
Forma de obtención de producto          0
Frecuencia de compra                    0
Precio                                  0
Cantidad comprada (unidad de medida)    0
Lugar de compra                         0
Codigo de producto                      0
dtype: int64

Cantidad total de valores ausente

Se evidencia que se ha elimiado los valores nulos en el precio en base al codigo de producto 110091. Lo que se ha realizado es reemplazar el promedio por codigo de producto en los valores nulos o vacios en este caso.

In [ ]:
df_final.shape

(1759681, 10)

In [ ]:
# Filtrar registros donde 'Precio' no es nulo o vacío
df_final = df_final[df_final['Precio'].notna() & (df_final['Precio'] != '')]

In [ ]:
# Lista de columnas a verificar
cols_to_fill = [
    'Frecuencia de compra',
    'Cantidad comprada (unidad de medida)',
    'Lugar de compra'
]

# Para cada columna, calcular la moda por 'Codigo de producto' y reemplazar valores nulos o vacíos
for col in cols_to_fill:
    # Calcular la moda por 'Codigo de producto'
    mode_by_product = df_final.groupby('Codigo de producto')[col].apply(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

    # Reemplazar valores nulos o vacíos con la moda calculada
    df_final[col] = df_final.groupby('Codigo de producto')[col].transform(lambda x: x.fillna(mode_by_product[x.name]))

    # Reemplazar valores vacíos con la moda calculada (en caso de que la columna sea de tipo string)
    df_final[col] = df_final.groupby('Codigo de producto')[col].transform(lambda x: x.replace('', mode_by_product[x.name]))


In [ ]:
df_final.shape

(1759159, 10)

In [ ]:
df_final.head()

CONGLOME  VIVIENDA  HOGAR        Nombre del producto  \
1       5030         8     11      Pan Corriente o común   
8       5030         8     11  TOSTADAS INTEGRALES A GRA   
16      5030         8     11   Arroz Corriente a Granel   
28      5030         8     11                Papa Blanca   
29      5030         8     11              Papa Amarilla   

   Forma de obtención de producto Frecuencia de compra  Precio  \
1                        Comprado                    2  150.00   
8                        Comprado                    4    0.37   
16                       Comprado                    2    1.00   
28                       Comprado                    3    1.00   
29                       Comprado                    3    0.50   

   Cantidad comprada (unidad de medida) Lugar de compra  Codigo de producto  
1                                     2               6              110091  
8                                     9               4              110087  
16                                    9               6              110100  
28                                    9               6              110915  
29                                    9               6              110913

In [ ]:
# Identificar valores nulos
null_values = df_final.isnull().sum()

# Identificar valores vacíos (asumiendo que las columnas con valores de tipo string pueden tener valores vacíos)
empty_values = df_final[df_final == ''].sum()

# Sumar ambos resultados para obtener el total de valores nulos o vacíos por columna
total_missing = null_values + empty_values

# Imprimir el resultado
print(total_missing)

CONGLOME                                0.0
VIVIENDA                                0.0
HOGAR                                   0.0
Nombre del producto                       0
Forma de obtención de producto            0
Frecuencia de compra                      0
Precio                                  0.0
Cantidad comprada (unidad de medida)      0
Lugar de compra                           0
Codigo de producto                      0.0
dtype: object


In [ ]:
# Identificar valores nulos
null_values = df_final.isnull().sum()

# Identificar valores vacíos en columnas de tipo object (string)
empty_values = df_final[df_final.select_dtypes(include=['object']) == ''].sum()

# Sumar ambos resultados para obtener el total de valores nulos o vacíos por columna
total_missing = null_values + empty_values

# Imprimir el resultado
print(total_missing)

CONGLOME                                0.0
VIVIENDA                                0.0
HOGAR                                   0.0
Nombre del producto                       0
Forma de obtención de producto            0
Frecuencia de compra                      0
Precio                                  0.0
Cantidad comprada (unidad de medida)      0
Lugar de compra                           0
Codigo de producto                      0.0
dtype: object


In [ ]:
df_final.shape

(1759159, 10)

In [ ]:
# Verificar una muestra de los datos que parecen estar vacíos en la columna 'Frecuencia de compra'
print(df_final[df_final['Frecuencia de compra'] == '']['Frecuencia de compra'].head())

# Reemplazar espacios en blanco (y otros caracteres no visibles) con NaN
df_final.replace(r'^\s*$', np.nan, regex=True, inplace=True)

Series([], Name: Frecuencia de compra, dtype: object)


In [ ]:
# Lista de columnas a verificar
cols_to_fill = [
    'Frecuencia de compra',
    'Cantidad comprada (unidad de medida)',
    'Lugar de compra'
]

# Para cada columna, calcular la moda por 'Codigo de producto' y reemplazar valores nulos o vacíos
for col in cols_to_fill:
    # Calcular la moda por 'Codigo de producto', excluyendo valores nulos y vacíos
    mode_by_product = df_final[df_final[col].notna() & (df_final[col] != '')].groupby('Codigo de producto')[col].apply(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

    # Reemplazar valores nulos o vacíos con la moda calculada
    df_final[col] = df_final.groupby('Codigo de producto')[col].transform(lambda x: x.replace('', np.nan).fillna(mode_by_product[x.name]))

In [ ]:
df_final.head(10)

CONGLOME  VIVIENDA  HOGAR        Nombre del producto  \
1       5030         8     11      Pan Corriente o común   
8       5030         8     11  TOSTADAS INTEGRALES A GRA   
16      5030         8     11   Arroz Corriente a Granel   
28      5030         8     11                Papa Blanca   
29      5030         8     11              Papa Amarilla   
35      5030         8     11               Azucar Rubia   
37      5030         8     11  Huevos a Granel de Gallin   
40      5030         8     11  HUEVO DE GALLINA DE CORRA   
52      5030         8     11           Pollo Eviscerado   
59      5030         8     11        Menudencia de Pollo   

   Forma de obtención de producto Frecuencia de compra      Precio  \
1                        Comprado                    2  150.000000   
8                        Comprado                    4    0.370000   
16                       Comprado                    2    1.000000   
28                       Comprado                    3    1.000000   
29                       Comprado                    3    0.500000   
35                       Comprado                    3    2.000000   
37                       Comprado                    3    0.360000   
40                           Otro                    4  200.760105   
52                       Comprado                   10    0.500000   
59                       Comprado                    3    0.250000   

   Cantidad comprada (unidad de medida) Lugar de compra  Codigo de producto  
1                                     2               6              110091  
8                                     9               4              110087  
16                                    9               6              110100  
28                                    9               6              110915  
29                                    9               6              110913  
35                                    9               6              111004  
37                                    9               6              110526  
40                                    2              11              110534  
52                                    9               6              110235  
59                                    9               6              110237

In [ ]:
df_final.to_csv('Dataset/Gastos_final.csv', index=False)